datafusion/optimizer/src/simplify_expressions/simplify_exprs.rs

```diff
@@ -59,13 +60,12 @@ impl SimplifyExpressions {
         plan: &LogicalPlan,
         execution_props: &ExecutionProps,
     ) -> Result<LogicalPlan> {
-        // We need to pass down the all schemas within the plan tree to `optimize_expr` in order to
-        // to evaluate expression types. For example, a projection plan's schema will only include
-        // projected columns. With just the projected schema, it's not possible to infer types for
-        // expressions that references non-projected columns within the same project plan or its
-        // children plans.
-        let info = plan
-            .all_schemas()
+        // Pass down the `children merge schema` and `plan schema` to evaluate expression types.
+        // pass all `child schema` and `plan schema` isn't enough, because like `t1 semi join t2 on
+        // on t1.id = t2.id`, each individual schema can't contain all the columns in it.
+        let children_merge_schema = DFSchemaRef::new(merge_schema(plan.inputs()));
```

Comment:
In some ways, it seems to me that we should *only* be using the children's schemas as the expressions *within* he `LogicalPlan` should be in terms of the plan's inputs (the children's schemas) not the plan's output (`plan.schema()`)

Reply:
> it seems to me that we should only be using the children's schemas.

I can't agree with it more.

But current code get attribution of some Expression by get it from `plan.schema()` instead of inferring/computing them from children ouput.

For example, some test will fail like `csv_query_group_by_and_having_and_where`.

Reply:
Something else to work on over time, perhaps